tests/spec/README.txt

```diff
@@ -0,0 +1,6 @@
+These tests specify the expected behaviour of ehrQL, and by extension, the QM.
```

Comment:
I think that the biggest problem with having these tests and unit tests like the one @iaindillingham wrote recently is that it's not clear to a developer where they should add new tests.

One mitigation I can see is that if this were used to generate spec document then "am I changing the language" or "should this example appear in the doc" could be used to inform that decision.

Reply:
Consider the following to be wrapped in a `with peters_opinion_on_a_wednesday():` context manager:

If there's any doubt about whether an ehrQL query will produce the right thing, we need to write an integration test.  The fast backend is plenty fast enough for this.

If we have an integration test, we don't also need a unit test for the same ehrQL query, unless the code in `query_language.py` becomes too convoluted.  At the moment, `query_language.py` contains almost no indirection and is about a simple as possible.

Checking that all binary operations work both ways feels like something that could belong either in the unit tests or in functional tests, depending on whether we think that there's any risk that engines will treat binary operations differently depending on which way round they are.

Do we need to have general team chat about this?

Reply:
Since you've put a lot of thought (and time) into this I'm happy for you to map out how it should work -- as long there is an unambiguous description that makes sense to the rest of the team. Whether that takes a meeting or just circulating a description for comment I don't mind. There is story coming up to document this in `DEVELOPERS.md`.